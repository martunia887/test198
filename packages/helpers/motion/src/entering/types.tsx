/**
 * Direction the motion is going.
 */
export type Direction = 'entering' | 'exiting';

/**
 * Direction an element can appear from
 */
export type From = 'top' | 'right' | 'bottom' | 'left';

/**
 * Common props all entering motions should make available for consumers.
 */
export interface MotionProps<TProps extends {}> {
  /**
   * Duration in `ms`.
   * How long the motion will take.
   */
  duration?: number;

  /**
   * Will callback when the motion has finished in the particular direction.
   * If it finished entering direction will be `entering`.
   * And vice versa for `exiting`.
   */
  onFinish?: (direction: Direction) => void;

  /**
   * Children as `function`.
   * Will be passed `props` for you to hook up.
   * The `direction` arg can be used to know if the motion is `entering` or `exiting`.
   */
  children: (opts: TProps, direction: Direction) => React.ReactNode;
}
