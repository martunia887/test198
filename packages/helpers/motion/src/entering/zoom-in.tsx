import React from 'react';
import { ObjectInterpolation } from '@emotion/core';
import EnteringMotion, { EnteringMotionProps } from './motion';

export const zoomInAnimation = (): ObjectInterpolation<undefined> => ({
  '0%': {
    opacity: 0,
    transform: 'scale(0.5)',
  },
  '50%': {
    opacity: 1,
  },
  '75%': {
    transform: 'scale(1.25)',
  },
  '100%': {
    transform: 'scale(1)',
  },
});

export const shrinkOutAnimation = (): ObjectInterpolation<undefined> => ({
  to: {
    opacity: 0,
    transform: 'scale(0.75)',
  },
});

const ZoomIn: React.FC<EnteringMotionProps> = ({
  children,
  duration = 125,
}: EnteringMotionProps) => {
  return (
    <EnteringMotion
      duration={duration}
      enteringAnimation={zoomInAnimation()}
      exitingAnimation={shrinkOutAnimation()}
      animationTimingFunction={() => 'ease-in-out'}
    >
      {children}
    </EnteringMotion>
  );
};

export default ZoomIn;
