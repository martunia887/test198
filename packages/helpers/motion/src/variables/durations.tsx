export const smallDurationMs = 100;

export const mediumDurationMs = 350;

export const largeDurationMs = 700;
