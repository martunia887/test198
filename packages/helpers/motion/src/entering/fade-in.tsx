import React from 'react';
import { ObjectInterpolation } from '@emotion/core';
import { easeInOut } from '../utils/curves';
import { largeDurationMs } from '../utils/durations';
import EnteringMotion, { EnteringMotionProps } from './motion';

export const fadeInAnimation = (): ObjectInterpolation<undefined> => ({
  from: {
    opacity: 0,
    transform: 'translate3d(0, 10%, 0)',
  },
  '50%': {
    opacity: 1,
  },
  to: {
    transform: 'none',
  },
});

export const fadeOutAnimation = (): ObjectInterpolation<undefined> => ({
  from: {
    opacity: 1,
    transform: 'translate3d(0, 0, 0)',
  },
  to: {
    opacity: 0,
    transform: 'translate3d(0, -10%, 0)',
  },
});

const FadeIn: React.FC<EnteringMotionProps> = ({
  children,
  duration = largeDurationMs,
}: EnteringMotionProps) => {
  return (
    <EnteringMotion
      duration={duration}
      enteringAnimation={fadeInAnimation()}
      exitingAnimation={fadeOutAnimation()}
      animationTimingFunction={() => easeInOut}
    >
      {children}
    </EnteringMotion>
  );
};

export default FadeIn;
