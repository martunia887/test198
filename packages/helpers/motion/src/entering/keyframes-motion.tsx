import React, { Ref, useEffect } from 'react';
import { ClassNames, keyframes, ObjectInterpolation } from '@emotion/core';

import { prefersReducedMotion } from '../utils/accessibility';
import { largeDurationMs } from '../utils/durations';
import { useSetTimeout } from '../utils/timer-hooks';

import { useExitingPersistence } from './exiting-persistence';
import { useStaggeredEntrance } from './staggered-entrance';
import { MotionProps, Direction } from './types';

/**
 * These are props that motions should use as their external props for consumers.
 * See [FadeIn](packages/helpers/motion/src/entering/fade-in.tsx) for an example usage.
 */
export interface KeyframesMotionProps
  extends MotionProps<{ className: string; ref: Ref<any> }> {
  /**
   * Can be used to pause the animation before it has finished.
   */
  isPaused?: boolean;
}

interface InternalKeyframesMotionProps extends KeyframesMotionProps {
  /**
   * Timing function to be used with the animation.
   * Receives the `direction` and expects a `string` return value.
   * Useful if you want a different curve when entering vs. exiting.
   */
  animationTimingFunction: (direction: Direction) => string;

  /**
   * CSS keyframes for the entering animation.
   */
  enteringAnimation: ObjectInterpolation<undefined>;

  /**
   * CSS keyframes for the exiting animation.
   */
  exitingAnimation?: ObjectInterpolation<undefined>;

  /**
   * Duration in `ms`.
   * How long the motion will take.
   */
  duration: number;
}

/**
 * Used to multipy the initial duration for exiting motions.
 */
const EXITING_MOTION_MULTIPLIER = 0.5;

/**
 * This is the base INTERNAL component used for all other entering motions.
 * This does not need Javascript to execute on the client so it will run immediately
 * for any SSR rendered React apps before the JS has executed.
 */
const EnteringMotion: React.FC<InternalKeyframesMotionProps> = ({
  children,
  animationTimingFunction,
  enteringAnimation,
  exitingAnimation,
  isPaused,
  onFinish: onFinishMotion,
  duration = largeDurationMs,
}: InternalKeyframesMotionProps) => {
  const staggered = useStaggeredEntrance();
  const {
    isExiting,
    onFinish: onExitFinished,
    appear,
  } = useExitingPersistence();
  const setTimeout = useSetTimeout();
  const paused = isPaused || !staggered.isReady;
  const delay = isExiting ? 0 : staggered.delay;
  const direction = isExiting ? 'exiting' : 'entering';
  const actualDuration = appear === false ? 0 : duration;

  useEffect(() => {
    if (paused) {
      return;
    }

    setTimeout(
      () => {
        if (direction === 'exiting') {
          onExitFinished && onExitFinished();
        }

        onFinishMotion && onFinishMotion(direction);
      },
      isExiting
        ? actualDuration * EXITING_MOTION_MULTIPLIER
        : actualDuration + delay,
    );
    // We ignore this for onFinishMotion as consumers could potentially inline the function
    // which would then trigger this effect every re-render.
    // We want to make it easier for consumers so we go down this path unfortunately.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    onExitFinished,
    direction,
    isExiting,
    actualDuration,
    delay,
    paused,
    setTimeout,
  ]);

  return (
    <ClassNames>
      {({ css }) =>
        children(
          {
            ref: staggered.ref,
            className: css({
              animationName: `${keyframes(
                isExiting
                  ? exitingAnimation || enteringAnimation
                  : enteringAnimation,
              )}`,
              animationTimingFunction: animationTimingFunction(direction),
              animationDelay: `${delay}ms`,
              animationFillMode: isExiting ? 'forwards' : 'backwards',
              animationDuration: `${
                isExiting
                  ? actualDuration * EXITING_MOTION_MULTIPLIER
                  : actualDuration
              }ms`,
              animationPlayState: paused ? 'paused' : 'running',
              ...prefersReducedMotion(),
            }),
          },
          direction,
        )
      }
    </ClassNames>
  );
};

export default EnteringMotion;
