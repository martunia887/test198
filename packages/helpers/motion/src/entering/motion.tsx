import React, { Ref, useEffect } from 'react';
import { ClassNames, keyframes, ObjectInterpolation } from '@emotion/core';
import { useStaggeredEntrance } from './staggered-entrance';
import { useExitingPersistence } from './exiting-persistence';
import { largeDurationMs } from '../utils/durations';
import { prefersReducedMotion } from '../utils/accessibility';

export type Direction = 'entering' | 'exiting';

/**
 * These are props that motions should use as their external props for consumers.
 * See [FadeIn](packages/helpers/motion/src/entering/fade-in.tsx) for an example usage.
 */
export interface EnteringMotionProps {
  /**
   * Duration in `ms`.
   * How long the motion will take.
   */
  duration?: number;

  /**
   * Use to pause the animation.
   */
  isPaused?: boolean;

  /**
   * Will callback when the motion has finished in the particular direction.
   * If it finished entering direction will be `entering`.
   * And vice versa for `exiting`.
   */
  onFinish?: (direction: Direction) => void;

  /**
   * Children as `function`.
   * Will be passed `props` for you to hook up.
   * The `direction` arg can be used to know if the motion is `entering` or `exiting`.
   */
  children: (
    props: { className: string; ref: Ref<any> },
    direction: Direction,
  ) => JSX.Element;
}

interface InternalEnteringMotionProps extends EnteringMotionProps {
  /**
   * Timing function to be used with the animation.
   * Receives the `direction` and expects a `string` return value.
   * Useful if you want a different curve when entering vs. exiting.
   */
  animationTimingFunction: (direction: Direction) => string;

  /**
   * CSS keyframes for the entering animation.
   */
  enteringAnimation: ObjectInterpolation<undefined>;

  /**
   * CSS keyframes for the exiting animation.
   */
  exitingAnimation?: ObjectInterpolation<undefined>;

  /**
   * Duration in `ms`.
   * How long the motion will take.
   */
  duration: number;

  /**
   * Children as function.
   * Will be passed `props` for you to hook up.
   * The `direction` arg can be used to know if the motion is entering or exiting.
   */
  children: (
    props: { className: string; ref: Ref<any> },
    direction: Direction,
  ) => JSX.Element;
}

/**
 * This is the base INTERNAL component used for all other entering motions.
 * This does not need Javascript to execute on the client so it will run immediately
 * for any SSR rendered React apps before the JS has executed.
 */
const EnteringMotion: React.FC<InternalEnteringMotionProps> = ({
  children,
  animationTimingFunction,
  enteringAnimation,
  exitingAnimation,
  isPaused,
  onFinish: onFinishMotion,
  duration = largeDurationMs,
}: InternalEnteringMotionProps) => {
  const staggered = useStaggeredEntrance();
  const { isExiting, onFinish: onExitFinished } = useExitingPersistence();
  const delay = isExiting ? 0 : staggered.delay;
  const direction = isExiting ? 'exiting' : 'entering';

  useEffect(() => {
    if (isPaused) {
      return;
    }

    const timeoutId = setTimeout(
      () => {
        if (direction === 'exiting') {
          onExitFinished && onExitFinished();
        }

        onFinishMotion && onFinishMotion(direction);
      },
      isExiting ? duration * 0.5 : duration + delay,
    );

    return () => {
      clearTimeout(timeoutId);
    };
  }, [
    onExitFinished,
    onFinishMotion,
    direction,
    isExiting,
    duration,
    delay,
    isPaused,
  ]);

  return (
    <ClassNames>
      {({ css }) =>
        children(
          {
            ref: staggered.ref,
            className: css({
              animationName: `${keyframes(
                isExiting
                  ? exitingAnimation || enteringAnimation
                  : enteringAnimation,
              )}`,
              animationTimingFunction: animationTimingFunction(direction),
              animationDelay: `${delay}ms`,
              animationFillMode: isExiting ? 'forwards' : 'backwards',
              animationDuration: `${isExiting ? duration * 0.5 : duration}ms`,
              animationPlayState:
                staggered.isReady || !isPaused ? 'running' : 'paused',
              ...prefersReducedMotion(),
            }),
          },
          direction,
        )
      }
    </ClassNames>
  );
};

export default EnteringMotion;
