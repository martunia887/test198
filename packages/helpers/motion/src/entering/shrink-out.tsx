import React, { useLayoutEffect } from 'react';
import { useExitingPersistence } from './exiting-persistence';
import { MotionProps } from './types';
import { useElementRef } from '../utils/use-element-ref';
import { useRequestAnimationFrame, useSetTimeout } from '../utils/timer-hooks';
import { smallDurationMs } from '../utils/durations';
import { easeIn } from '../utils/curves';

interface ShrinkOutProps extends MotionProps<{ ref: React.Ref<any> }> {}

const ShrinkOut: React.FC<ShrinkOutProps> = ({
  children,
  duration = smallDurationMs,
  onFinish,
}: ShrinkOutProps): any => {
  const [element, setElementRef] = useElementRef();
  const exiting = useExitingPersistence();
  const requestAnimationFrame = useRequestAnimationFrame();
  const setTimeout = useSetTimeout();
  const direction = exiting.isExiting ? 'exiting' : 'entering';

  useLayoutEffect(() => {
    if (exiting.isExiting && element) {
      element.style.width = `${element.offsetWidth}px`;
      element.style.boxSizing = 'border-box';

      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          element.style.transitionTimingFunction = easeIn;
          element.style.transitionDuration = `${duration}ms`;
          element.style.transitionProperty = 'width,margin';
          element.style.width = '0px';
          element.style.margin = '0px';

          setTimeout(() => {
            exiting.onFinish && exiting.onFinish();
            onFinish && onFinish('exiting');
          }, duration);
        });
      });
    }
  });

  return children({ ref: setElementRef }, direction);
};

export default ShrinkOut;
