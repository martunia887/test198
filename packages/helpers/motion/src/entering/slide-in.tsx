import React from 'react';
import { ObjectInterpolation } from '@emotion/core';
import { easeOut, easeIn } from '../utils/curves';
import { mediumDurationMs } from '../utils/durations';
import EnteringMotion, { EnteringMotionProps, Direction } from './motion';

export type From = 'top' | 'right' | 'bottom' | 'left';

export const slideInAnimation = (
  from: From,
  direction: Direction,
): ObjectInterpolation<undefined> => {
  const initial = direction === 'entering' ? '0%' : '100%';
  const end = direction === 'entering' ? '100%' : '0%';
  const fromMap = {
    top: 'translate3d(0, -100%, 0)',
    right: 'translate3d(100%, 0, 0)',
    bottom: 'translate3d(0, 100%, 0)',
    left: 'translate3d(-100%, 0, 0)',
  };

  return {
    [initial]: {
      transform: fromMap[from],
    },
    [end]: {
      transform: 'none',
    },
  };
};

interface SlideInProps extends EnteringMotionProps {
  /**
   * Direction the element will slide in from.
   * E.g. `"right"` will slide in from the right to the left.
   */
  from: From;
}

const SlideIn: React.FC<SlideInProps> = ({
  children,
  from,
  duration = mediumDurationMs,
}: SlideInProps) => {
  return (
    <EnteringMotion
      duration={duration}
      enteringAnimation={slideInAnimation(from, 'entering')}
      exitingAnimation={slideInAnimation(from, 'exiting')}
      animationTimingFunction={direction =>
        direction === 'entering' ? easeOut : easeIn
      }
    >
      {children}
    </EnteringMotion>
  );
};

export default SlideIn;
