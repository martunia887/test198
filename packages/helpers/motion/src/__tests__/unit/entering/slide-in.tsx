import React from 'react';
import { render } from '@testing-library/react';
import SlideIn, { slideInAnimation } from '../../../entering/slide-in';
import ExitingPersistence from '../../../entering/exiting-persistence';
import { easeOut, easeIn } from '../../../utils/curves';
import { mediumDurationMs } from '../../../utils/durations';
import { Direction, From } from '../../../entering/types';

jest.mock('../../../utils/accessibility');

describe('<SlideIn />', () => {
  it('should default to medium duration', () => {
    const { getByTestId } = render(
      <SlideIn from="left">
        {props => <div data-testid="target" {...props} />}
      </SlideIn>,
    );

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-duration',
      `${mediumDurationMs}ms`,
    );
  });

  it('should override default duration', () => {
    const { getByTestId } = render(
      <SlideIn duration={123} from="left">
        {props => <div data-testid="target" {...props} />}
      </SlideIn>,
    );

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-duration',
      '123ms',
    );
  });

  it('should slide in easing out', () => {
    const { getByTestId } = render(
      <SlideIn from="left">
        {props => <div data-testid="target" {...props} />}
      </SlideIn>,
    );

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-timing-function',
      easeOut,
    );
  });

  it('should slide out easing in', () => {
    const { getByTestId, rerender } = render(
      <ExitingPersistence>
        <SlideIn from="left">
          {props => <div data-testid="target" {...props} />}
        </SlideIn>
      </ExitingPersistence>,
    );

    rerender(<ExitingPersistence>{false}</ExitingPersistence>);

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-timing-function',
      easeIn,
    );
  });

  ['entering', 'exiting'].forEach(direction => {
    ['top', 'right', 'bottom', 'left'].forEach(from => {
      it(`should animate ${direction} from ${from}`, () => {
        expect(
          slideInAnimation(from as From, direction as Direction),
        ).toMatchSnapshot();
      });
    });
  });
});
