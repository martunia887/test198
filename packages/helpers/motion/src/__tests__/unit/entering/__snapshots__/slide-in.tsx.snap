// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlideIn /> should animate entering from bottom 1`] = `
Object {
  "0%": Object {
    "transform": "translate3d(0, 100%, 0)",
  },
  "100%": Object {
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from left 1`] = `
Object {
  "0%": Object {
    "transform": "translate3d(-100%, 0, 0)",
  },
  "100%": Object {
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from right 1`] = `
Object {
  "0%": Object {
    "transform": "translate3d(100%, 0, 0)",
  },
  "100%": Object {
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from top 1`] = `
Object {
  "0%": Object {
    "transform": "translate3d(0, -100%, 0)",
  },
  "100%": Object {
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate exiting from bottom 1`] = `
Object {
  "0%": Object {
    "transform": "none",
  },
  "100%": Object {
    "transform": "translate3d(0, 100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from left 1`] = `
Object {
  "0%": Object {
    "transform": "none",
  },
  "100%": Object {
    "transform": "translate3d(-100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from right 1`] = `
Object {
  "0%": Object {
    "transform": "none",
  },
  "100%": Object {
    "transform": "translate3d(100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from top 1`] = `
Object {
  "0%": Object {
    "transform": "none",
  },
  "100%": Object {
    "transform": "translate3d(0, -100%, 0)",
  },
}
`;
